### PR TITLE
Add support for make child without fork

### PR DESCRIPTION
**Pull request by jessehui**

### CODE REVIEW

File: jobs.c [hunk 1]
```diff
@@ -5134,3 +5138,565 @@ restore_pgrp_pipe (p)
...
+       as fd 0, but have not changed the buffered stream corresponding to
+       the old fd 0.  We don't want to sync the stream in this case. */
+    if (default_buffered_input != -1 &&
+            (!async_p || default_buffered_input > 0)) {
+        sync_buffered_stream (default_buffered_input);
+    }
+#endif /* BUFFERED_INPUT */
+
+    int child_pid = 0;
+    struct nofork_child_args *arg = (struct nofork_child_args *)calloc(1,
```
Comment (guzongmin) [inline jobs.c:5645]: check the arg to confirm it is not NULL

File: jobs.c [hunk 2]
```diff
@@ -5134,3 +5138,565 @@ restore_pgrp_pipe (p)
...
+    } else if (strcmp(operation, "subshell") == 0) {
+      child_pid = children_routine_for_subshell(arg);
+    } else if (strcmp(operation, "process_subst") == 0) {
+      arg->open_for_read_in_child = open_for_read_in_child;
+      child_pid = children_routine_for_process_subst(arg);
+    } else if (strcmp(operation, "cmd_subst") == 0) {
+      child_pid = children_routine_for_cmd_subst(arg);
+    } else {
+      sys_error("unhandled operation type: %s\n", operation);
+      return -1;
```
Comment (guzongmin) [inline jobs.c:5673]: free arg before return

Comment (jessehui) [reply jobs.c:5673]: Done. Thanks for your review.

File: jobs.c (hunk 1 above)
Comment (jessehui) [reply jobs.c:5645]: Done.

File: jobs.c [hunk 3]
```diff
@@ -5134,3 +5138,602 @@ restore_pgrp_pipe (p)
...
+            sys_error("posix_spawn_file_actions_addclose failed: %d", errno);
+            return ret;
+        }
+    }
+    return ret;
+}
+
+// Create a temp scrip file for Bash to execute. In this way, Bash can spawn self and use this script
+// name as an argument.
+static int create_temp_script(char *subshell_script, char *command) {
```
Comment (jessehui) [inline jobs.c:5224]: Maybe adding a file lock here just to be safer.

Comment (jessehui) [reply jobs.c:5224]: Done.

**Pull request merged**